akka-actor/src/main/scala/akka/dispatch/affinity/AffinityPool.scala

```diff
@@ -0,0 +1,217 @@
+    locked(mainLock) {
+      val worker = new ThreadPoolWorker(q)
+      workers.add(worker)
+      worker.startWorker()
+    }
+  }
+
+  private def tryEnqueue(command: Runnable) = {
+    // does a mod on the hash code to determine which queue to put task in
+    val queueIndex = Math.abs(command.hashCode()) & (numThreads - 1)
+    workQueues(queueIndex).add(command)
+  }
+
+  private def onWorkerExit(w: ThreadPoolWorker): Unit = {
+    locked(mainLock) {
+      workers.remove(w)
+    }
+  }
+
+  private def onWorkerFailure(w: ThreadPoolWorker): Unit = {
+    locked(mainLock) {
+      workers.remove(w)
+      addWorker(workers, w.q)
+    }
+  }
+
+  private class ThreadPoolWorker(val q: BoundedTaskQueue) extends Runnable {
+
+    sealed trait WorkerState
+    case object NotStarted extends WorkerState
+    case object InExecution extends WorkerState
+    case object Idle extends WorkerState
+
+    private val thread = tf.newThread(this)
+    @volatile var workerState: WorkerState = NotStarted
+
+    def startWorker() = thread.start()
+
+    def runCommand(command: Runnable) = {
+      workerState = InExecution
+      try
+        command.run()
+      catch {
+        case x: RuntimeException ⇒ throw x
+        case e: Error            ⇒ throw e
+        case t: Throwable        ⇒ throw new Error(t)
+      } finally {
+        workerState = Idle
+      }
+    }
+
+    override def run(): Unit =
+      Try {
+        workerState = Idle
+        while (poolState == Running || !q.isEmpty) {
+          val c = q.poll()
+          if (c != null) runCommand(c)
+        }
+      } match {
+        case Failure(_) ⇒ onWorkerFailure(this)
+        case _          ⇒ onWorkerExit(this)
+      }
+
+    def interruptIfStarted() =
+      if (workerState != NotStarted && !thread.isInterrupted)
+        thread.interrupt()
+
+    def interruptIfIdle() =
+      if (workerState == Idle && !thread.isInterrupted)
+        thread.interrupt()
+
+  }
+
+  override def execute(command: Runnable): Unit = {
+    if (command == null)
+      throw new NullPointerException
+    if (!(poolState == Running && tryEnqueue(command)))
+      reject(command)
+  }
+
+  private def reject(command: Runnable) = throw new RejectedExecutionException("Task " + command.toString + " rejected from " + this.toString)
+
+  //TODO: Implement via wait condition
+  override def awaitTermination(timeout: Long, unit: TimeUnit): Boolean = true
+
+  // attempts to transition pool into Terminated state
+  // TODO: need to implement this so we are sure all workers are gone and there are no more tasks left in the queue
+  private def attemptPoolTermination() = poolState = Terminated
+
+  override def shutdownNow(): util.List[Runnable] = {
+    poolState = ShuttingDown
+    //interrupts all workers
+    locked(mainLock) {
+      workers.foreach(_.interruptIfStarted())
+    }
+    poolState = ShutDown
+    seqAsJavaList(workQueues.flatMap(_.drain))
```

Comment:
This is not safe as AbstractBoundedNodeQueue is a Single Consumer queue and the Worker is consuming the queue concurrently with this thread (the one which executes this logic).